src/SourcesCommand.cc

```diff
@@ -130,6 +141,197 @@ struct DirExistsCache {
+  }
+
+  string gdb_script_host_path = resource_path() + "bin/rr-gdb-script-host.py";
+  pid_t pid;
+  char* gdb_script_host_argv[] = {
+    strdup(gdb_script_host_path.c_str()),
+    strdup(output_file.name.c_str()),
+    strdup(gdb_script.c_str()),
+    strdup(program.c_str()),
+    NULL,
+  };
+  ret = posix_spawn(&pid, gdb_script_host_path.c_str(), &file_actions, NULL,
+                    gdb_script_host_argv, environ);
+  if (ret != 0) {
+    FATAL() << "posix_spawn failed with " << ret;
+  }
+
+  // Ignore the return values during cleanup.
+  posix_spawn_file_actions_destroy(&file_actions);
+
+  close(stdin_pipe_fds[0]);
+
+  this->pid = pid;
+  this->pipe_fd = stdin_pipe_fds[1];
+  this->output_file = fopen(output_file.name.c_str(), "r");
+  if (!this->output_file) {
+    FATAL() << "Failed to open gdb script host result file " << output_file.name;
+  }
+
+  synchronize();
+}
+
+DebugDirs DebugDirManager::process_one_binary(const string& binary_path) {
+  if (pipe_fd < 0) {
+    return DebugDirs();
+  }
+
+  auto len = binary_path.length();
+  size_t written = write(pipe_fd, binary_path.c_str(), len);
+  if (written != len) {
+    FATAL() << "Failed to write filename";
+  }
+  written = write(pipe_fd, "\n", 1);
+  if (written != 1) {
+    FATAL() << "Failed to write trailing newline";
+  }
+
+  synchronize();
+  return read_result();
+}
+
+DebugDirs DebugDirManager::read_result() {
+  char buf[4096];
+  DebugDirs result;
+  size_t index;
+  const char delimiter[2] = ":";
+
+  if (pipe_fd < 0) {
+    return result;
+  }
+
+  if (!fgets(buf, sizeof(buf) - 1, output_file)) {
+    FATAL() << "Failed to read gdb script output";
+  }
+  index = strcspn(buf, "\n");
+  buf[index] = 0;
+
+  char* token = strtok(buf, delimiter);
+  while (token != NULL) {
+    char* buf = realpath(token, NULL);
+    if (buf) {
+      auto s = string(buf);
+      result.debug_file_directories.push_back(s);
+      LOG(debug) << "gdb script added debug dir '" << s << "'";
+      free(buf);
+    } else {
+      LOG(debug) << "realpath(" << token << ") = " << strerror(errno);
+    }
+    token = strtok(NULL, delimiter);
+  }
+
+  if (!fgets(buf, sizeof(buf) - 1, output_file)) {
+    FATAL() << "Failed to read gdb script output";
+  }
+  index = strcspn(buf, "\n");
+  buf[index] = 0;
+
+  token = strtok(buf, delimiter);
+  while (token != NULL) {
+    char* buf = realpath(token, NULL);
+    if (buf) {
+      auto s = string(buf);
+      result.source_directories.push_back(s);
+      LOG(debug) << "gdb script added source dir '" << s << "'";
+      free(buf);
+    } else {
+      LOG(debug) << "realpath(" << token << ") = " << strerror(errno);
+    }
+    token = strtok(NULL, delimiter);
+  }
+
+  return result;
+}
+
+void DebugDirManager::synchronize() {
+  int status;
+  if (waitpid(pid, &status, WUNTRACED) == -1) {
+    FATAL() << "Failed to wait on gdb script host";
+  }
+
+  if (!(WIFSTOPPED(status) && WSTOPSIG(status) == SIGSTOP)) {
```

Comment:
This is weird! why can't we just communicate over the pipe like a regular program?

Reply:
I'm a bit afraid this will interact poorly with ctrl-Z job control

Reply:
Mmm yeah. I was in the mindset of writing kernel behavior tests :P

Reply:
Are you going to change this? I don't like it :-)